spec.html

```diff
@@ -35464,7 +35665,18 @@ <h1>Notation</h1>
         <p>The descriptions below use the following internal data structures:</p>
         <ul>
           <li>
-            A <dfn id="pattern-charset" variants="CharSets">CharSet</dfn> is a mathematical set of characters. In the context of a Unicode pattern, “all characters” means the CharSet containing all code point values; otherwise “all characters” means the CharSet containing all code unit values.
+            A <dfn>CharSetElement</dfn> is one of the two following entities:
+            <ul>
+              <li>
+                If _rer_.[[UnicodeSets]] is *false*, then a CharSetElement is a character in the sense of the Pattern Semantics above.
+              </li>
+              <li>
+                If _rer_.[[UnicodeSets]] is *true*, then a CharSetElement is a sequence whose elements are characters in the sense of the Pattern Semantics above. This includes the empty sequence, sequences of one character, and sequences of more than one character.
```

Comment:
This means that whenever [[UnicodeSets]] is true, a CharSet contains _sequences_ of characters, not individual characters.

But there's a number of places which operate on CharSets as if they contain characters, without first guarding on "not [[UnicodeSets]]". Among them:

- AllCharacters returns a CharSet containing code points, not sequences of code points
  - Quite a few other places construct a CharSet from individual code points
- CharacterSetMatcher has "If there exists a member _a_ of _A_ such that Canonicalize(_rer_, _a_) [...]", where _A_ in this context is a CharSet. Canonicalize's second argument is a single character, not a sequence of characters.

The first kind can be worked around by adding some extra explanation here:

```suggestion
                If _rer_.[[UnicodeSets]] is *true*, then a CharSetElement is a sequence whose elements are characters in the sense of the Pattern Semantics above. This includes the empty sequence, sequences of one character, and sequences of more than one character. For convenience, when working with CharSetElements of this kind, individual characters are treated interchangably with sequences of one character.
```

That doesn't fix the type error in CharacterSetMatcher, though. But I believe the argument to CharacterSetMatcher actually is guaranteed to contain only length-1 sequences, so it could be fixed by adding an assert to that effect (which I've suggested below).

Reply:
Suggestion applied in https://github.com/mathiasbynens/ecma262/pull/27

Reply:
re the added sentence, I think putting it in the singular would be clearer: "... an individual character is treated interchangably with a sequence of one character"

Reply:
Done